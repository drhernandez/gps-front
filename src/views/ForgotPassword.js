import React from "react";
import { Link, withRouter } from "react-router-dom";
import {
  Container,
  Row,
  Col,
  Card,
  CardBody,
  FormInput,
  FormGroup,
  FormFeedback,
  Form,
  Button,
  Alert
} from "shards-react";
import validations from "../utils/ValidationsUtil";
import constants from "../utils/Constants";
import "../styles/login.css";
var _ = require('lodash');

const errorsDefault = {
  email: {
    required: false,
    valid: false
  }
}

class ForgotPassword extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errors: errorsDefault,
      remember: false,
      snackbar: {
        visible: false,
        type: constants.Themes.SUCCESS
      }
    }

    this.restorePassword.bind(this);
    this.invalidateError.bind(this);
  };

  restorePassword(e) {
    e.preventDefault()
    e.persist();
    const errors = _.clone(errorsDefault);
    errors.email.required = !validations.validateRequired(e.target.email.value);
    errors.email.valid = !validations.validateEmail(e.target.email.value);
    
    if (Object.values(errors).find(fieldValidations => Object.values(fieldValidations).find(value => value))) {  // Si alguno de los errores está en true...
      this.setState({
        errors: errors
      })
    } else {
      //call service
      const sent = true
      if (sent) {
        this.setState({
          snackbar: {
            visible: true,
            type: constants.Themes.SUCCESS
          }
        })
        setTimeout(() => {
          this.props.history.push("/signin");
        }, 4000);
      } else {
        this.setState({
          snackbar: {
            visible: true,
            type: constants.Themes.ERROR
          }
        })
        setTimeout(() => {
          this.setState({
            snackbar: {
              visible: false,
              type: constants.Themes.SUCCESS
            }
          })
        }, 4000);
      }
    }
  }
  
  invalidateError(field) {
    let errors = this.state.errors;
    Object.keys(errors[field]).forEach(validationType => errors[field][validationType] = false)
    this.setState({
      errors: errors
    })
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4 h-100">
        <Row noGutters className="h-100">
          <Col className="mx-auto login">
            <Card small>
              <CardBody className="px-4">
                <h5 className="login__titulo text-center mt-5 mb-4">Recupera tu contraseña</h5>
                <Form onSubmit={e => this.restorePassword(e)} noValidate>
                  <FormGroup>
                    <label htmlFor="email">Email</label>
                    <FormInput
                      type="email"
                      id="email"
                      placeholder="Dirección de correo"
                      autoComplete="email"
                      invalid={this.state.errors.email.required || this.state.errors.email.valid}
                      onChange={() => this.invalidateError("email")}
                    />
                    {this.state.errors.email.required && <FormFeedback>Campo requerido</FormFeedback>}
                    {this.state.errors.email.valid && <FormFeedback>La dirección de correo no es válida. Una dirección válida se vería así: micorreo@dominio.com</FormFeedback>}
                    <div className="px-2">
                      <small>Recibirás un email con las instrucciones para recuperar tu contraseña.</small>
                    </div>
                  </FormGroup>
                  <Button type="submit" pill className="d-table mx-auto mb-4">Recuperar contraseña</Button>
                  <FormGroup>
                    <Alert className="mb-3" open={this.state.snackbar.visible} theme={this.state.snackbar.type}>
                      {this.state.snackbar.type === constants.Themes.SUCCESS && "Ya te enviamos las instrucciones para recuperar tu contraseña, revisa tu casilla de correo. Te redireccionaremos al login"}
                      {this.state.snackbar.type === constants.Themes.ERROR && "Algo salió mal al intentar enviarte las instrucciones. Inténtelo de nuevo."}
                    </Alert>
                  </FormGroup>
                </Form>
              </CardBody>
            </Card>
            <div className="px-2 pt-3 login__center_links">
              <Link to="/signin">Volver al login</Link>
            </div>
          </Col>
        </Row>
      </Container>
    );
  }
}

export default withRouter(ForgotPassword);