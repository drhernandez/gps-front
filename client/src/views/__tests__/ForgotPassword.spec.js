import React from 'react';
import renderer from 'react-test-renderer';
import { Router } from "react-router-dom";
import { createMemoryHistory } from 'history'
import { shallow, mount } from 'enzyme';

jest.mock("../../api/services/recoveryService", () => ({ generateToken: jest.fn() }));
import RecoveryService from '../../api/services/recoveryService';

import ForgotPassword from "../ForgotPassword"

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

describe('ForgotPassword view', () => {
  beforeEach(async () => {
    jest.resetModules();
    jest.useFakeTimers()

    await RecoveryService.generateToken.mockImplementation(() => {
      return Promise.resolve();
    });
  });

  it("snapshot renders", () => {
    const history = createMemoryHistory()
    const component = renderer.create(
      <Router history={history}>
        <ForgotPassword />
      </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("Link Volver al login should redirect to login", () => {
    const wrapper = shallow(
      <ForgotPassword.WrappedComponent />
    );

    expect(wrapper.find('Link').at(0).props().to).toBe('/signin')
  });

  it("Should redirect to login when button is pressed", async () => {
    const history = createMemoryHistory({
      initialEntries: ['/forgot-password']
    });
    const wrapper = mount(
      <Router history={history}>
        <ForgotPassword />
      </Router>
    );

    const mockedEvent = {
      target: {
        'email': {
          value: 'test@test.com'
        }
      },
      preventDefault: () => { },
      persist: () => { }
    }

    expect(history.location.pathname).toBe("/forgot-password")

    wrapper.find('ForgotPassword').instance().restorePassword(mockedEvent);
    await flushPromises();
    jest.runAllTimers();

    expect(history.location.pathname).toBe("/signin")
  });
});