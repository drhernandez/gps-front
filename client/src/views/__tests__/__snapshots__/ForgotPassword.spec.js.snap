// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword view snapshot renders 1`] = `
<div
  className="main-content-container px-4 h-100 container-fluid"
>
  <div
    className="h-100 no-gutters row"
  >
    <div
      className="mx-auto login col"
    >
      <div
        className="card card-small"
      >
        <div
          className="px-4 card-body"
        >
          <h5
            className="login__titulo text-center mt-5 mb-4"
          >
            Recupera tu contraseña
          </h5>
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="email"
              >
                Email
              </label>
              <input
                autoComplete="email"
                className="form-control"
                id="email"
                onChange={[Function]}
                placeholder="Dirección de correo"
                type="email"
              />
              <div
                className="px-2"
              >
                <small>
                  Recibirás un email con las instrucciones para recuperar tu contraseña.
                </small>
              </div>
            </div>
            <button
              className="d-table mx-auto mb-4 btn btn-accent"
              onClick={[Function]}
              type="submit"
            >
              Recuperar contraseña
            </button>
            <div
              className="form-group"
            />
          </form>
        </div>
      </div>
      <div
        className="px-2 pt-3 login__center_links"
      >
        <a
          href="/signin"
          onClick={[Function]}
        >
          Volver al login
        </a>
      </div>
    </div>
  </div>
</div>
`;
